src/urllib3/response.py

```diff
@@ -472,11 +494,11 @@ def tell(self) -> int:
         """
         return self._fp_bytes_read
 
-    def _init_length(self, request_method):
+    def _init_length(self, request_method: Optional[str]) -> Optional[int]:
         """
         Set initial length value for Response content if available.
         """
-        length = self.headers.get("content-length")
+        length: Optional[Union[int, str]] = self.headers.get("content-length")
```

Comment:
Having this typed here as `Optional[Union[int, str]]` instead of `Optional[str]` is causing a `str()` call below to be necessary. Let's have this variable be named `content_length` and keep `length` as it is elsewhere but define it above as `length: Optional[int]`?

Reply:
Then we have `UnboundLocalError: local variable 'length' referenced before assignment` error in 
https://github.com/urllib3/urllib3/blob/2ec06d1185f1642dfe8ff767a9ad3541bd0ddc39/src/urllib3/response.py#L525

when `content-length` is `None`.

Reply:
I was going to try to explain in text but it's way faster to just copy-paste and make the edit:

```python
length: Optional[int]
content_length: Optional[str] = self.headers.get("content-length")

if content_length is not None:
    if self.chunked:
        # This Response will fail with an IncompleteRead if it can't be
        # received as chunked. This method falls back to attempt reading
        # the response before raising an exception.
        log.warning(
            "Received response with both Content-Length and "
            "Transfer-Encoding set. This is expressly forbidden "
            "by RFC 7230 sec 3.3.2. Ignoring Content-Length and "
            "attempting to process response as Transfer-Encoding: "
            "chunked."
        )
        return None

    try:
        # RFC 7230 section 3.3.2 specifies multiple content lengths can
        # be sent in a single Content-Length header
        # (e.g. Content-Length: 42, 42). This line ensures the values
        # are all valid ints and that as long as the `set` length is 1,
        # all values are the same. Otherwise, the header is invalid.
        lengths = {int(val) for val in content_length.split(",")}
        if len(lengths) > 1:
            raise InvalidHeader(
                "Content-Length contained multiple "
                "unmatching values (%s)" % content_length
            )
        length = lengths.pop()
    except ValueError:
        length = None
    else:
        if length < 0:
            length = None

else:  # if content_length is None
    length = None

# Convert status to int for comparison
# In some cases, httplib returns a status of "_UNKNOWN"
try:
    status = int(self.status)
except ValueError:
    status = 0

# Check for responses that shouldn't include a body
if status in (204, 304) or 100 <= status < 200 or request_method == "HEAD":
    length = 0

return length
```

Basically the change is the `else: length=None`, does this work?

Reply:
Thanks @sethmlarson, I've replaced your cone block